### PR TITLE
adding stream interceptor for logging middleware

### LINKED ISSUE #3351: Kratos middleware seems to only apply to unary gRPC operations
#### What happened:
We've started using Kratos middleware (recovery, logging, validate) in our project, and noticed that, on the gRPC side, they only seem to apply to unary operations. Our streaming operations don't get validated or logged, and a panic takes down the server (versus on a unary endpoint, they're validated and logged as expected, and panics are handled.)

#### What you expected to happen:
We expected to either have the middleware apply seamlessly to streaming requests or to have separate streaming middleware we can register (which, to be fair, may exist, but we weren't able to find it.)

#### How to reproduce it (as minimally and precisely as possible):
Repro project: https://github.com/wscalf/streaming-repro

- "SayHello" is the normal hello world operation + a 50/50 chance of panicking to test the recovery middleware.
- "KeepSayingHello" is an added operation that streams the response every second until the client terminates the connection, at which point it panics (intentionally, to test 
Commands:
- `grpcurl -plaintext -d '{"name":""}' localhost:9000 helloworld.v1.Greeter.SayHello`: gets correctly rejected with a validation error
- `grpcurl -plaintext -d '{"name":""}' localhost:9000 helloworld.v1.Greeter.KeepSayingHello`: should get rejected (it's using the same message) but does not. Also, is coded to panic when the client closes, which should trigger the recovery middleware, but it doesn't- it just exits
- `grpcurl -plaintext -d '{"name":"Kratos"}' localhost:9000 helloworld.v1.Greeter.SayHello`: is valid, and is coded to have a 50/50 chance of panicking, which does correctly trigger the recovery middleware, returning an error to the client and not exiting.

#### Environment:
- Kratos version (use `kratos -v`): kratos version v2.7.3
- Go version (use `go version`): go version go1.22.2 linux/amd64
- OS (e.g: `cat /etc/os-release`):
```
NAME="Fedora Linux"
VERSION="40 (Workstation Edition)"
ID=fedora
VERSION_ID=40
VERSION_CODENAME=""
PLATFORM_ID="platform:f40"
PRETTY_NAME="Fedora Linux 40 (Workstation Edition)"
ANSI_COLOR="0;38;2;60;110;180"
LOGO=fedora-logo-icon
CPE_NAME="cpe:/o:fedoraproject:fedora:40"
DEFAULT_HOSTNAME="fedora"
HOME_URL="https://fedoraproject.org/"
DOCUMENTATION_URL="https://docs.fedoraproject.org/en-US/fedora/f40/system-administrators-guide/"
SUPPORT_URL="https://ask.fedoraproject.org/"
BUG_REPORT_URL="https://bugzilla.redhat.com/"
REDHAT_BUGZILLA_PRODUCT="Fedora"
REDHAT_BUGZILLA_PRODUCT_VERSION=40
REDHAT_SUPPORT_PRODUCT="Fedora"
REDHAT_SUPPORT_PRODUCT_VERSION=40
SUPPORT_END=2025-05-13
VARIANT="Workstation Edition"
VARIANT_ID=workstation

```
- Others:

### PR DESCRIPTION
**Pull request by akoserwal**
#### Description (what this PR does / why we need it):

Update the stream interceptor to handle stream use-case

#### Which issue(s) this PR fixes (resolves / be part of):

fixes #3351

### CODE REVIEW

**Comment by xuyang2** [comment]
似乎可以直接用 `"github.com/grpc-ecosystem/go-grpc-middleware/v2/interceptors/logging"`

- https://pkg.go.dev/github.com/grpc-ecosystem/go-grpc-middleware/v2/interceptors/logging
- https://github.com/grpc-ecosystem/go-grpc-middleware/tree/main/interceptors/logging/examples
- https://github.com/grpc-ecosystem/go-grpc-middleware/tree/main/interceptors/logging

**Comment by akoserwal** [comment]
@shenqidebaozi I updated the PR to have a middleware abstraction for the streaming middleware. grpc-gateway middleware also configure the unary and stream as separate interceptors configs
https://github.com/grpc-ecosystem/go-grpc-middleware/blob/main/examples/server/main.go#L136

**Comment by akoserwal** [comment]
@go-kratos/contributor please take a look and provide feedback. Another proposal to unify the unary and stream interface: https://github.com/akoserwal/kratos/pull/1

**Comment by shenqidebaozi** [comment]
ok， Let's discuss this design

File: middleware/middleware.go [hunk 1]
```diff
@@ -2,14 +2,21 @@ package middleware
...
 import (
 	"context"
+	"google.golang.org/grpc"
 )
 
 // Handler defines the handler invoked by Middleware.
 type Handler func(ctx context.Context, req interface{}) (interface{}, error)
 
+// StreamHandler defines the handler invoked by Middleware for stream calls.
+type StreamHandler func(srv interface{}, stream grpc.ServerStream) error
```
Comment (shenqidebaozi) [inline middleware/middleware.go:12]: Can I directly use the middleware interface? For example, putting the stream into req

File: middleware/logging/logging_stream.go [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+)
+
+type loggingServerStream struct {
+	req any
+	grpc.ServerStream
+	logger log.Logger
+}
+
+// StreamServer is a server logging middleware for gRPC streams.
+func StreamServer(logger log.Logger) middleware.StreamMiddleware {
```
Comment (shenqidebaozi) [inline middleware/logging/logging_stream.go:19]: Now we need to develop a streaming system for all middleware

File: middleware/logging/logging_stream.go [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+				kind      string
+				operation string
+			)
+			ctx := stream.Context()
+			startTime := time.Now()
+			if info, ok := transport.FromClientContext(ctx); ok {
+				kind = info.Kind().String()
+				operation = info.Operation()
+			}
+			wrappedStream := &loggingServerStream{
```
Comment (shenqidebaozi) [inline middleware/logging/logging_stream.go:34]: Can we treat middleware as a processor for every streaming call

File: middleware/middleware.go (hunk 1 above)
Comment (akoserwal) [reply middleware/middleware.go:12]: @wscalf has another proposal to unify the middleware interface: https://github.com/akoserwal/kratos/pull/1

Comment (wscalf) [reply middleware/middleware.go:12]: It's mostly scratchwork I wrote to check assumptions/make sure the interfaces lined up and in no way ready for review (I was holding off on spending a lot of time with it unless the Kratos team were interested in seeing more), but it goes along with what I was suggesting here: https://github.com/go-kratos/kratos/issues/3351#issuecomment-2223919198

**Comment by akoserwal** [comment]
@shenqidebaozi updated the interceptor based on your suggestions

File: middleware/middleware.go [hunk 4]
```diff
@@ -2,14 +2,21 @@ package middleware
...
+
 // Middleware is HTTP/gRPC transport middleware.
 type Middleware func(Handler) Handler
 
+// StreamMiddleware is gRPC stream transport middleware.
+type StreamMiddleware func(StreamHandler) StreamHandler
```
Comment (shenqidebaozi) [inline middleware/middleware.go:18]: Can all of these be deleted

Comment (akoserwal) [reply middleware/middleware.go:18]: removed

Comment (shenqidebaozi) [reply middleware/middleware.go]: It looks much simpler now, can you add some individual tests?

Comment (akoserwal) [reply middleware/middleware.go]: sure, I'll add tests

File: transport/grpc/interceptor.go [hunk 5]
```diff
@@ -2,6 +2,7 @@ package grpc
 
 import (
 	"context"
+	"github.com/go-kratos/kratos/v2/internal/matcher"
```
Comment (shenqidebaozi) [inline transport/grpc/interceptor.go:5]: fix  import sort

File: transport/grpc/interceptor.go [hunk 6]
```diff
@@ -85,3 +99,45 @@ func (s *Server) streamServerInterceptor() grpc.StreamServerInterceptor {
...
+	return ctx.Value(stream{}).(grpc.ServerStream)
+}
+
+func (w *wrappedStream) SendMsg(m interface{}) error {
+	err := w.ServerStream.SendMsg(m)
+	info, _ := transport.FromServerContext(w.ctx)
+
+	h := func(ctx context.Context, req interface{}) (interface{}, error) {
+		return req, nil
+	}
```
Comment (shenqidebaozi) [inline transport/grpc/interceptor.go:118]: It seems that the send or recv data should be included here

File: transport/grpc/interceptor.go [hunk 7]
```diff
@@ -76,7 +79,18 @@ func (s *Server) streamServerInterceptor() grpc.StreamServerInterceptor {
...
 
-		ws := NewWrappedStream(ctx, ss)
+		h := func(ctx context.Context, req interface{}) (interface{}, error) {
+			ctx = context.WithValue(ctx, stream{
+				ServerStream: ss,
+				middlware:    s.middleware,
+			}, ss)
+			return handler(srv, ss), nil
+		}
+		if next := s.middleware.Match(info.FullMethod); len(next) > 0 {
```
Comment (shenqidebaozi) [inline transport/grpc/interceptor.go:89]: Additionally, I think we should add a configuration item to specify which middleware is used for streaming send recv

Comment (akoserwal) [reply transport/grpc/interceptor.go:89]: What would you recommend?

Comment (shenqidebaozi) [reply transport/grpc/interceptor.go:90]: grpc.StreamMiddleware

Comment (akoserwal) [reply transport/grpc/interceptor.go:90]: Do you mean `grpc.ServerStream`? There is no `grpc.StreamMiddleware`

**Comment by akoserwal** [comment]
@shenqidebaozi
@tonybase
@daemon365

Please review

File: middleware/middleware.go [hunk 8]
```diff
@@ -19,3 +19,5 @@ func Chain(m ...Middleware) Middleware {
 		return next
 	}
 }
+
+type StreamMiddleware func(Handler) Handler
```
Comment (shenqidebaozi) [inline middleware/middleware.go:23]: It doesn't seem to be used

**Comment by shenqidebaozi** [comment]
It seems that we still lack Client Stream Middleware

File: middleware/middleware.go (hunk 8 above)
Comment (akoserwal) [reply middleware/middleware.go:23]: I will remove it

**Review by shenqidebaozi — DISMISSED** [review]

**Comment by shenqidebaozi** [comment]
@go-kratos/contributor PTAL.

File: transport/grpc/client.go [hunk 9]
```diff
@@ -6,6 +6,8 @@ import (
 	"fmt"
 	"time"
 
+	"github.com/go-kratos/kratos/v2/internal/matcher"
```
Comment (demoManito) [inline transport/grpc/client.go:9]: image

File: transport/grpc/interceptor.go [hunk 10]
```diff
@@ -85,3 +101,48 @@ func (s *Server) streamServerInterceptor() grpc.StreamServerInterceptor {
...
+	grpc.ServerStream
+	streamMiddleware matcher.Matcher
+}
+
+func GetStream(ctx context.Context) grpc.ServerStream {
+	return ctx.Value(stream{}).(grpc.ServerStream)
+}
+
+func (w *wrappedStream) SendMsg(m interface{}) error {
+	h := func(ctx context.Context, req interface{}) (interface{}, error) {
```
Comment (demoManito) [inline transport/grpc/interceptor.go:115]: _ context.Context

File: transport/grpc/interceptor.go [hunk 11]
```diff
@@ -85,3 +101,48 @@ func (s *Server) streamServerInterceptor() grpc.StreamServerInterceptor {
...
+	if next := w.middleware.Match(info.Operation()); len(next) > 0 {
+		h = middleware.Chain(next...)(h)
+	}
+
+	_, err := h(w.ctx, m)
+	return err
+}
+
+func (w *wrappedStream) RecvMsg(m interface{}) error {
+	h := func(ctx context.Context, req interface{}) (interface{}, error) {
```
Comment (demoManito) [inline transport/grpc/interceptor.go:133]: ditto

File: transport/grpc/server_test.go [hunk 12]
```diff
@@ -11,6 +11,8 @@ import (
 	"testing"
 	"time"
 
+	"google.golang.org/grpc/metadata"
+
 	"google.golang.org/grpc"
```
Comment (demoManito) [inline transport/grpc/server_test.go:16]: No blank lines are required

File: transport/grpc/server_test.go [hunk 13]
```diff
@@ -280,6 +282,82 @@ func TestServer_unaryServerInterceptor(t *testing.T) {
...
+		streamMiddleware: matcher.New(),
+	}
+
+	srv.streamMiddleware.Use(EmptyMiddleware())
+
+	mockStream := &mockServerStream{
+		ctx: srv.baseCtx,
+	}
+
+	handler := func(srv interface{}, stream grpc.ServerStream) error {
```
Comment (demoManito) [inline transport/grpc/server_test.go:340]: _ interface{}

**Review by demoManito — APPROVED** [review]

File: middleware/middleware.go (hunk 8 above)
Comment (shenqidebaozi) [reply middleware/middleware.go:23]: It seems like it hasn't been deleted yet

File: transport/grpc/client.go [hunk 14]
```diff
@@ -166,7 +167,7 @@ func dial(ctx context.Context, insecure bool, opts ...ClientOption) (*grpc.Clien
 		unaryClientInterceptor(options.middleware, options.timeout, options.filters),
 	}
 	sints := []grpc.StreamClientInterceptor{
-		streamClientInterceptor(options.filters),
+		streamClientInterceptor(options.middleware, options.filters),
```
Comment (shenqidebaozi) [inline transport/grpc/client.go:170]: streamMiddleware？

**Comment by shenqidebaozi** [comment]
@akoserwal I think it's ready now, but I need to test it again

**Review by shenqidebaozi — APPROVED** [review]

**Pull request merged**